### PR TITLE
feat: Add simulator as additional metadata of OS metadata in user agent

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

https://github.com/awslabs/aws-sdk-swift/issues/874

## Description of changes

Using compiler directive, add "simulator" as additional metadata to OS metadata in user agent header if SDK is being used within simulators.

Also, make the `value` of additional metadata optional to follow the SEP.

The change was tested using a simple iOS application & confirming that user agent header header is changed as expected:
```
User-Agent: aws-sdk-swift/1.0 ua/2.0 api/s3#1.0 os/ios#17.4.0 md/simulator lang/swift#5.10 cfg/retry-mode#legacy
```

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/Core/AWSClientRuntime/UserAgent/ExecutionEnvMetadata.swift [hunk 1]
```diff
@@ -18,6 +18,9 @@ extension ExecutionEnvMetadata: CustomStringConvertible {
     }
 
     static func detectExecEnv() -> ExecutionEnvMetadata? {
+        #if targetEnvironment(simulator)
+        return ExecutionEnvMetadata(name: "simulator")
```
Comment (jbelkins) [inline Sources/Core/AWSClientRuntime/UserAgent/ExecutionEnvMetadata.swift:22]: From the user-agent SEP:
> Execution environment name can be retrieved from the environment variable AWS_EXECUTION_ENV as defined in this [SEP](https://code.amazon.com/packages/AwsDrSeps/blobs/ca26ae803530fb10429a218c1c0a3e25ee07cceb/--/seps/accepted/shared/user_agent.md). The SDK MUST include this in the header if present.

I don't think simulator is the intended use of this UA field, and I'm afraid we will mess up data collection if we put other values into this field.

**Review by jbelkins — CHANGES_REQUESTED** [review]

File: Sources/Core/AWSClientRuntime/UserAgent/ExecutionEnvMetadata.swift (hunk 1 above)
Comment (sichanyoo) [reply Sources/Core/AWSClientRuntime/UserAgent/ExecutionEnvMetadata.swift:22]: That's fair, I think what I could do is append this as additional metadata to OS metadata then.

File: Sources/Core/AWSClientRuntime/UserAgent/AdditionalMetadata.swift [hunk 2]
```diff
@@ -9,12 +9,21 @@ import Foundation
```
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/UserAgent/AdditionalMetadata.swift:9]: The `value` is optional according to SEP. Changed accordingly.

File: Sources/Core/AWSClientRuntime/UserAgent/OSMetadata.swift [hunk 3]
```diff
@@ -11,11 +11,18 @@ struct OSMetadata {
     let family: PlatformOperatingSystem
```
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/UserAgent/OSMetadata.swift:11]: If compiled for simulator target, add `md/simulator` additional metadata.

File: Tests/Core/AWSClientRuntimeTests/UserAgent/AWSUserAgentMetadataTests.swift [hunk 4]
```diff
@@ -17,6 +17,15 @@ class AWSUseragentMetadataTests: XCTestCase {
     let executionEnvMetadata = ExecutionEnvMetadata(name: "e123")
```
Comment (sichanyoo) [inline Tests/Core/AWSClientRuntimeTests/UserAgent/AWSUserAgentMetadataTests.swift:17]: Add a test for simulator case.

**Review by jbelkins — APPROVED** [review]
Approved after consulting with SDK data team about how to implement.

**Pull request merged**